### PR TITLE
p2p/simulations: Refactor

### PR DESCRIPTION
**Pull request by lmars**
This is a grab bag of refactoring to get this ready for merging into master (apologies for the big diff), summary:

* Move `p2p/adapters` -> `p2p/simulations/adapters` since the adapters should only be used in simulations
* Remove unused `p2p/adapters/reporter.go`
* Run `gofmt` on some files (this happens automatically when I save files)
* Rename the existing `NodeAdapter` interface to `Node` and introduce a new `NodeAdapter` interface which represents objects capable of starting nodes (i.e. `SimNodeAdapter`, `DockerAdapter` etc.)
* Initialise `simulations.Network` with a `NodeAdapter` which it uses to create nodes
* `NodeConfig` now contains a `Service` field which indicates which service to run on the node
* `p2p/simulations/examples` has ben enhanced to run a simple ping-pong protocol
* `p2p/testing` now uses a "mock" service for triggering / expecting messages which avoids the need to access the underlying `p2p.MsgReadWriter` directly, and also removes the need for synchronising through the old `Connc` / `Readyc` channels of `adapters.Peer`
* `ProtocolSession.TestDisconnected` now waits for peer events to check for disconnects rather than relying on `Peer.Errc` (which has been removed)

### CODE REVIEW

File: p2p/simulations/network.go [hunk 1]
```diff
@@ -510,62 +502,6 @@ func (self *NodeEvent) ToControlEvent() *NodeControlEvent {
...
-	pubkey := crypto.FromECDSAPub(&key.PublicKey)
-	copy(id[:], pubkey[1:])
-	return &NodeConfig{
-		Id:         &adapters.NodeId{NodeID: id},
-		PrivateKey: key,
-	}
-}
-
 // TODO: ignored for now
 type QueryConfig struct {
```
Comment (zelig) [inline p2p/simulations/network.go]: can we remove these? thx

Comment (lmars) [reply p2p/simulations/network.go]: Done.

**Comment by lmars** [comment]
@zelig I have also added message events in 6b97141.

File: p2p/peer.go [hunk 2]
```diff
@@ -82,12 +83,13 @@ const (
 )
 
 // PeerEvent is an event emitted when peers are either added or dropped from
-// a p2p.Server
+// a p2p.Server or when a message is sent or received on a peer connection
 type PeerEvent struct {
```
Comment (zelig) [inline p2p/peer.go:87]: I think we will need more info of the message and not sure we necessarily care about the Code and size. 
What is we just optionally pass a `func(interface{}) ([]byte, bool) ` function to the eventer which would rlp-encode any 
data we wanna map the message to for different use cases.

Of course this is perfect for now as is.

**Review by zelig — APPROVED** [review]

File: p2p/peer.go (hunk 2 above)
Comment (lmars) [reply p2p/peer.go:87]: I think we should focus on writing the simulations that require these message events and then modify when we have a better idea of the requirements.

Comment (lmars) [reply p2p/peer.go:87]: Also, these events are being emitted via RPC, so it isn't straight forward to provide a callback like you suggest.

Comment (zelig) [reply p2p/peer.go:87]: sure i meant it should be via the service setup of course not real time

Comment (zelig) [reply p2p/peer.go:87]: > I think we should focus on writing the simulations that require these message events and then modify when we have a better idea of the requirements.

agreed hence no a change request, just to record

**Pull request merged**